### ISSUE TITLE
🚀 Separate configuration file and multiple profiles

### ISSUE THREAD

**Issue by ShogunPanda** [dandavison/delta#1301]
Hello all!
First of all, thanks for this awesome tool which I love every day.
In very humble opinion there are two features that are missing:

## External configuration file

Rather than storing all options in `.gitconfig`, can we store the configuration inside an external file and just reference that from the .gitconfig?

Like:

```
# .gitconfig
[delta]
  config = /path/to/config
```

and 

```sh
delta --config=/path/to/config
```

I don't have any preference for the config format, it might be the same gitconfig format, TOML or YAML.

## Support multiple profiles

Following the proposal above (but given we use environment variable it is technically independent of it), we might add a `profile` option which specifies which section to load. For instance:

```
# .gitconfig

[delta]
  config = /path/to/config
  profile = main
``` 

and 

```sh
delta --config=/path/to/config --profile=main
```

will look for section `delta.main` rather than simply `delta`.

What do you think about it?

**Comment by dandavison** [comment]
Hi @ShogunPanda,

> Rather than storing all options in .gitconfig, can we store the configuration inside an external file and just reference that from the .gitconfig?

Good news! Yes, we can already do this, because gitconfig supports `[include.path](https://git-scm.com/docs/git-config#_includes)`. Here's an example from my own config files:

https://github.com/dandavison/dotfiles/blob/fea9c053d00b8c1e3318b04d8a3753967a8e7648/git/gitconfig#L34

https://github.com/dandavison/dotfiles/blob/fea9c053d00b8c1e3318b04d8a3753967a8e7648/git/delta.gitconfig

> Following the proposal above (but given we use environment variable it is technically independent of it), we might add a profile option which specifies which section to load.

Perhaps you could do this using https://git-scm.com/docs/git-config#_conditional_includes? If you try it out, it would be great if you could report back and let me know what you get working. Also, if you want to update the delta user manual with this info that would be great.

**Comment by ShogunPanda** [comment]
> Good news! Yes, we can already do this, because gitconfig supports [`include.path`](https://git-scm.com/docs/git-config#_includes). Here's an example from my own config files:

I see. But this does not allow to completely bypass the `.gitconfig` file (especially when using the CLI outside of GIT).

> Perhaps you could do this using https://git-scm.com/docs/git-config#_conditional_includes? If you try it out, it would be great if you could report back and let me know what you get working. Also, if you want to update the delta user manual with this info that would be great.

Same as above, while your suggestion probably work in the `.gitconfig` file, they are not flexible on CLI.

In case I would try to implement these feature by myself, would you accept a PR or you don't want these features to be implemented?

**Comment by dandavison** [comment]
What's an example of a situation when you wouldn't want to use gitconfig?

> In case I would try to implement these feature by myself, would you accept a PR or you don't want these features to be implemented?

I don't think we want to implement delta option parsing twice in two different file formats. There are some pretty complicated semantics involving delta "features". Wouldn't you say delta config is pretty tied to gitconfig at this point?

**Comment by ShogunPanda** [comment]
> What's an example of a situation when you wouldn't want to use gitconfig?

If, for instance, I use delta on a server where I have no git and I just want to diff files outside a GIT repo. In that case I'd like not to have to write a GIT config file but rather use a standalone one (where, without fear of future problems if the server is updated) I can only delta configuration only.


> I don't think we want to implement delta option parsing twice in two different file formats. There are some pretty complicated semantics involving delta "features". Wouldn't you say delta config is pretty tied to gitconfig at this point?

To be honest, the data format doesn't matter. We can definitely keep the `.gitconfig` format, so forget the YAML/JSON/whatever.

To summarize, what is left for me is the following two new features:

1. Being able, via CLI and/or environment variable, to read configuration from a different file different from `~/.gitconfig` (but written using the same syntax of `~/.gitconfig`
2. A CLI and/or environment variable to specify which section of the file to read. For instance, `delta -p secondary` would read the `delta.secondary` section in the configuration file. The default will be as it is today.

**Comment by dandavison** [comment]
> Being able, via CLI and/or environment variable, to read configuration from a different file different from `~/.gitconfig` (but written using the same syntax of `~/.gitconfig`

I think that this _should_ be possible already using the `GIT_CONFIG_GLOBAL` env var. However, libgit2 needs to support it. There is an open PR for that https://github.com/libgit2/libgit2/pull/6399 so perhaps it would help to help that PR along.

**Comment by ShogunPanda** [comment]
I'm not really familiar with Rust git2, but wouldn't [Config.open](https://docs.rs/git2/latest/git2/struct.Config.html#method.open) would allow us to open any file despite of the default loading strategy of libgit2?

**Comment by dandavison** [comment]
> wouldn't [Config.open](https://docs.rs/git2/latest/git2/struct.Config.html#method.open) would allow us to open any file despite of the default loading strategy of libgit2?

That sounds very plausible, i.e. that someone could implement a `--config-file` flag for delta. Whether someone does that or not, delta (i.e. libgit2) should support `GIT_CONFIG_GLOBAL` since delta is often used as git's pager and it is confusing for delta to not be using the same git config file as the git process that is invoking delta.

**Comment by ShogunPanda** [comment]
I see. If you're fine with it, I'll try to implement both `--config-file` and `--profile` CLI flags to start. Then eventually we can improve the PR with env variables.
Do we have a deal?

**Comment by dandavison** [comment]
Hi, OK, yes, I think `--config-file` makes sense. Thanks for offering to help out with delta!

I'm not sure about `--profile` yet. The concern that comes to mind is that that can already be done with `--features`. And "features" is already confusing enough for people so additionally having "profiles" might make it just too confusing.

**Comment by ShogunPanda** [comment]
Oh, I didn't know about features. It seems feature is exactly what I was looking for. I just have a question though:

Given the following `.gitconfig`:

```
[delta]
  line-numbers = true

[delta "whatever"]
  zero-style = dim syntax
```

Assuming that the default for `line-numbers` is false (which should be the case but if not it's just an example here), if I load the `whatever` feature will the default be loaded as well? 

If not, we are all set for this.

Otherwise then the only thing I see missing to implement the second part of my proposal is an option to skip loading the default.
In other words, I want to be able for a `feature` section to define the entire set of option without having to override the ones defined in the `[delta]` section.

**Comment by dandavison** [comment]
> Assuming that the default for `line-numbers` is false (which should be the case but if not it's just an example here), if I load the whatever feature will the default be loaded as well?

Yes, it will. `line-numbers` needs to have the value `true` or `false`, and it ends up with the default value if not overridden by a feature.

> If not, we are all set for this.

Hm, I'm not sure how the answer could have been "no".

There's a comment here documenting how option values are resolved which might be helpful: https://github.com/dandavison/delta/blob/master/src/options/get.rs#L11-L27

**Comment by ShogunPanda** [comment]
I see.
I'll play with the source code and eventually will ask for more insights. Thanks!
